modules/dnn/src/tensorflow/tf_importer.cpp

```diff
@@ -1113,6 +1115,74 @@ void TFImporter::parseReshape(tensorflow::GraphDef& net, const tensorflow::NodeD
     }
 }
 
+void TFImporter::parseExpandDims(tensorflow::GraphDef& net, const tensorflow::NodeDef& layer, LayerParams& layerParams)
+{
+    const std::string& name = layer.name();
+    const int num_inputs = layer.input_size();
+
+    CV_CheckGT(num_inputs, 0, "");
+    Pin inpId = parsePin(layer.input(0));
+    DataLayout inpLayout = getDataLayout(layer.input(0), data_layouts);
+
+    // Get input shape
+    std::vector<MatShape> inShape_, outShape_;
+    int inpIdindex = layer_id.find(inpId.name)->second;
+
+    dstNet.getLayerShapes(netInputShapes, inpIdindex, inShape_, outShape_);
+    MatShape inpShape = outShape_[0];
+    std::vector<int> outShape = inpShape;
+
+    int outShapeSize = outShape.size();
+
+    // 2nd blob is dims tensor
+    int axis = getConstBlob(layer, value_id, 1).int_val().Get(0);
+
+    if(axis == -1)
+    {
+        axis = inpShape.size();
+    }
+
+    CV_Assert(0 <= axis && axis <= inpShape.size());
+
+    if(axis == inpShape.size() && inpShape.size() >= 4)
+    {
+        int order[] = {0, 2, 3, 1};  // From OpenCV's NCHW to NHWC.
+        addPermuteLayer(order, name + "/nhwc", inpId);
+
+        // Convert shape From OpenCV's NCHW to NHWC.
+        if(inpLayout == DATA_LAYOUT_NHWC)
+        {
+            for(int i = 1; i < outShapeSize - 1; i++)
+            {
+                std::swap(outShape[i], outShape[i+1]);
+            }
+        }
+    }
+
+    if(inpLayout == DATA_LAYOUT_NHWC)
+    {
+        // toNCHW
+        axis = (axis != 0)?(axis % outShapeSize + 1):0;
```

Comment:
Can you leave the comment about how this works? And check that for all axis in [-5; 4] this works correctly and add this test to extra?

Reply:
Thank you for your code review, I will update your comment in the next version.